x-pack/plugins/saved_objects_tagging/server/tags/tags_client.ts

```diff
@@ -0,0 +1,65 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+import { SavedObjectsClientContract } from 'src/core/server';
+import { TagSavedObject, TagAttributes, ITagsClient } from '../../common/types';
+import { tagSavedObjectTypeName } from '../../common/constants';
+import { TagValidationError } from './errors';
+import { validateTag } from './validate_tag';
+import { savedObjectToTag } from './utils';
+
+interface TagsClientOptions {
+  client: SavedObjectsClientContract;
+}
+
+export class TagsClient implements ITagsClient {
+  private readonly soClient: SavedObjectsClientContract;
+  private readonly type = tagSavedObjectTypeName;
+
+  constructor({ client }: TagsClientOptions) {
+    this.soClient = client;
+  }
+
+  public async create(attributes: TagAttributes) {
+    const validation = validateTag(attributes);
+    if (!validation.valid) {
+      throw new TagValidationError('Error validating tag attributes', validation);
+    }
+    const raw = await this.soClient.create<TagAttributes>(this.type, attributes);
+    return savedObjectToTag(raw);
+  }
+
+  public async update(id: string, attributes: TagAttributes) {
+    const validation = validateTag(attributes);
+    if (!validation.valid) {
+      throw new TagValidationError('Error validating tag attributes', validation);
+    }
+    const raw = await this.soClient.update<TagAttributes>(this.type, id, attributes);
+    return savedObjectToTag(raw as TagSavedObject); // all attributes are updated, this is not a partial
+  }
+
+  public async get(id: string) {
+    const raw = await this.soClient.get<TagAttributes>(this.type, id);
+    return savedObjectToTag(raw);
+  }
+
+  public async getAll() {
+    const result = await this.soClient.find<TagAttributes>({
+      type: this.type,
+      perPage: 10000,
+    });
+
+    return result.saved_objects.map(savedObjectToTag);
+  }
+
+  public async delete(id: string) {
+    await this.soClient.delete(this.type, id);
+    // `removeReferencesTo` security check is the same as a `delete` operation's, so we can use the scoped client here.
+    // If that was to change, we would need to use the internal client instead. A FTR test is ensuring
+    // that this behave properly even with only 'tag' SO type write permission.
+    await this.soClient.removeReferencesTo(this.type, id);
```

Comment:
I think we should swap around the order here, first remove all references then remove the tag object. Otherwise if we get a conflict error from update_by_query (because someone saved an object with a tag reference) then the tag will be gone but some objects will still reference it.

This way the UI is also more consistent, so a user can see easily that the delete tag operation failed because the tag is still showing in the tags management UI (even though it has less references than before)

Reply:
Both orders contains issues, (without proper transaction rollback, but let's not troll again here). Reversing the order, would, in case of error during the first call, cause the tag to stay, while having some objects already unassociated from the tag. This sounds as impactful, or do we think this is better?

Really not sure about the correct order here, but I can reverse if we think this is preferable. @restrry, your opinion?

Reply:
If we leave it the way it is if an error happens to occur when removing references using an `update_by_query`, will the user be able to click "delete tag" again and have the dangling references removed?

Reply:
Nope. That's good enough for me, will switch.

Reply:
Done in https://github.com/elastic/kibana/pull/79096/commits/cb15dbb7d634390e6c6296d691e351e6ee7a83b3